internal/controller/controllers/agent_controller.go

```diff
@@ -204,6 +210,132 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, &h.Host, h.ClusterID, nil, false)
 }
 
+func findMatchingNode(agent *aiv1beta1.Agent, nodes []corev1.Node) *corev1.Node {
+	addresses := make([]string, 0)
+	for _, intf := range agent.Status.Inventory.Interfaces {
+		for _, ip := range append(intf.IPV4Addresses, intf.IPV6Addresses...) {
+			splitIp := strings.Split(ip, "/")
+			if len(splitIp) == 2 {
+				addresses = append(addresses, splitIp[0])
+			}
+		}
+	}
+	for i := range nodes {
+		n := &nodes[i]
+		if strings.EqualFold(n.Name, agent.Spec.Hostname) {
+			return n
+		}
+		for _, addr := range n.Status.Addresses {
+			if addr.Type == corev1.NodeInternalIP && funk.ContainsString(addresses, addr.Address) {
+				return n
+			}
+		}
+	}
+	return nil
+}
+
+func isNodeReady(node *corev1.Node) bool {
+	for _, c := range node.Status.Conditions {
+		if c.Status == corev1.ConditionTrue && c.Type == corev1.NodeReady {
+			return true
+		}
+	}
+	return false
+}
+
+func isNodeApproved(agent *aiv1beta1.Agent, nodes []corev1.Node) bool {
+	if agent.Status.DebugInfo.State == models.HostStatusAddedToExistingCluster &&
+		agent.Status.Progress.CurrentStage == models.HostStageRebooting {
+		node := findMatchingNode(agent, nodes)
+		if node != nil {
+			return isNodeReady(node)
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) shouldApproveCSRs(agent *aiv1beta1.Agent, clients ClusterKubeClient) (bool, error) {
+
+	nodes, err := clients.ListNodes()
+	if err != nil {
+		return false, err
+	}
+	return !isNodeApproved(agent, nodes.Items), nil
+}
+func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
+	for _, c := range csr.Status.Conditions {
+		if c.Type == certificatesv1.CertificateApproved {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) isAssociatedWithAgent(csr *certificatesv1.CertificateSigningRequest, agent *aiv1beta1.Agent) (bool, error) {
```

Comment:
I think we should do other checks on the CSR, as I've suggested elsewhere.

If possible, I think we should get as close as possible to the checks done by the cluster-machine-approver

https://github.com/openshift/cluster-machine-approver/blob/master/pkg/controller/csr_check.go


Maybe we can vendor `cluster-machine-approver` and reuse some of its internal functions? What do you think?

At a high level, we should at least check the IPs/SAN and make sure they match. Also, make sure that we were expecting a new CSR rather than just listing them all and checking them.

Reply:
Referring to https://github.com/openshift/assisted-service/pull/3169#discussion_r796648762 - OK
SAN is optional in CSR.  The above file is all lower-cased so it us package private.  It is used just by the controller.  In addition it looks for machines and do some logic on it which is irrelevant for us. 
Can take all checks from the above file.  SAN is not checked there (or maybe I missed it).  Not sure it is required to be there?
We are expecting CSR because we trigger CSR approval only when host moves to Done and the agent is not Done yet which we consider rebooting